blobstore-s3/src/lib.rs

```diff
@@ -0,0 +1,830 @@
+            .key(&arg.chunk.object_id)
+            .body(ByteStream::from(bytes))
+            .send()
+            .await
+        {
+            Ok(_) => Ok(PutObjectResponse::default()),
+            Err(e) => {
+                error!(
+                    "put_object: Bucket({}) Object({}): {}",
+                    &arg.chunk.container_id,
+                    &arg.chunk.object_id,
+                    &e.to_string(),
+                );
+                Err(RpcError::Other(e.to_string()))
+            }
+        }
+    }
+
+    /// Retrieve object from s3 storage.
+    async fn get_object(
+        &self,
+        ctx: &Context,
+        arg: &blobstore::GetObjectRequest,
+    ) -> RpcResult<GetObjectResponse> {
+        let max_chunk_size = self.max_chunk_size();
+        // If the object is not found, or not readable, get_object_metadata will return error.
+        let meta = self
+            .get_object_metadata(ctx, &arg.container_id, &arg.object_id)
+            .await?;
+        // calculate content_length requested, with error checking for range bounds
+        let bytes_requested = match (arg.range_start, arg.range_end) {
+            (None, Some(end)) => meta.content_length.min(end + 1),
+            (Some(start), None) if start < meta.content_length => meta.content_length - start,
+            (Some(start), Some(end)) if (start <= end) && start < meta.content_length => {
+                meta.content_length.min(end - start + 1)
+            }
+            (None, None) => meta.content_length,
+            _ => 0,
+        };
+        if bytes_requested == 0 {
+            return Ok(GetObjectResponse {
+                content_length: 0,
+                content_encoding: meta.content_encoding.clone(),
+                content_type: meta.content_type.clone(),
+                initial_chunk: Some(Chunk {
+                    bytes: vec![],
+                    container_id: arg.container_id.clone(),
+                    object_id: arg.object_id.clone(),
+                    is_last: true,
+                    offset: 0,
+                }),
+                success: true,
+                error: None,
+            });
+        }
+        let get_object_req = self
+            .0
+            .get_object()
+            .bucket(&arg.container_id)
+            .key(&arg.object_id)
+            .set_range(to_range_header(arg.range_start, arg.range_end));
+        match get_object_req.send().await {
+            Ok(mut object_output) => {
+                let len = object_output.content_length as u64;
+                if len > bytes_requested {
+                    // either the math is wrong above, or we misunderstood the api.
+                    // docs say content_length is "Size of the body in bytes"
+                    error!(
+                        "requested {} bytes but more bytes {} were returned!",
+                        bytes_requested, len
+                    );
+                }
```

Comment:
I'm curious if this can be affected by some kind of race condition, e.g. some object in S3 that is being continuously appended to. I think this is the right implementation though, we don't want to download something incorrect. to that end, should we be asserting that `len == bytes_requested`?

(edit: I see that if len < bytes_requested that means we need to stream chunks, so this is all good. Just leaving this comment in case others see it)

Reply:
S3 objects can't be appended to - they are fixed size.